crates/runtime/src/datafusion/cluster/codec/spice_logical_codec.rs

```diff
@@ -0,0 +1,111 @@
+use datafusion::common::{DataFusionError, Result, TableReference, exec_err};
+use datafusion::prelude::SessionContext;
+use datafusion_expr::registry::FunctionRegistry;
+use datafusion_expr::{Extension, LogicalPlan, ScalarUDF};
+use datafusion_proto::logical_plan::LogicalExtensionCodec;
+use std::fmt::Debug;
+use std::sync::Arc;
+use tokio::runtime::Handle;
+use tokio::task;
+
+/// Serialization support for custom Spice logical nodes
+pub struct SpiceLogicalCodec {
+    inner: Arc<dyn LogicalExtensionCodec>,
+    runtime: Option<Arc<Runtime>>,
+}
+
+impl Debug for SpiceLogicalCodec {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SpiceLogicalCodec")
+    }
+}
+
+impl SpiceLogicalCodec {
+    #[must_use]
+    pub fn new_with_runtime(runtime: Arc<Runtime>) -> Arc<dyn LogicalExtensionCodec> {
+        Arc::new(Self {
+            inner: Arc::new(BallistaLogicalExtensionCodec::default()),
+            runtime: Some(runtime),
+        })
+    }
+
+    #[must_use]
+    pub fn new_codec() -> Arc<dyn LogicalExtensionCodec> {
+        Arc::new(Self {
+            inner: Arc::new(BallistaLogicalExtensionCodec::default()),
+            runtime: None,
+        })
+    }
+
+    fn runtime(&self) -> Result<Arc<Runtime>> {
+        self.runtime.clone().ok_or(DataFusionError::Execution(
+            "SpiceLogicalCodec did not bind a Runtime handle. This is a bug.".to_string(),
+        ))
+    }
+}
+
+impl LogicalExtensionCodec for SpiceLogicalCodec {
+    fn try_decode(
+        &self,
+        buf: &[u8],
+        inputs: &[LogicalPlan],
+        ctx: &SessionContext,
+    ) -> Result<Extension> {
+        if let Ok(ext) = self.inner.try_decode(buf, inputs, ctx) {
+            return Ok(ext);
+        }
+
+        let name = serde_json::from_slice::<String>(buf)
+            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+
+        let node = match name.as_str() {
+            "BytesProcessedNode" => Extension {
+                node: Arc::new(BytesProcessedNode::new(inputs[0].clone())),
+            },
+            other => return exec_err!("SpiceLogicalCodec does not support {other}"),
+        };
+
+        Ok(node)
+    }
+
+    fn try_encode(&self, node: &Extension, buf: &mut Vec<u8>) -> Result<()> {
+        if let Ok(()) = self.inner.try_encode(node, buf) {
+            return Ok(());
+        }
+
+        let node_name = serde_json::to_vec(node.node.name())
+            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+        buf.extend_from_slice(&node_name[..]);
+        Ok(())
+    }
+
+    // Look up the table ref in the context instead of ser/de
+    fn try_decode_table_provider(
+        &self,
+        _buf: &[u8],
+        table_ref: &TableReference,
+        _schema: SchemaRef,
+        ctx: &SessionContext,
+    ) -> Result<Arc<dyn TableProvider>> {
+        task::block_in_place(|| Handle::current().block_on(ctx.table_provider(table_ref.clone())))
```

Comment:
It is possible to get this synchronously: https://github.com/spiceai/spiceai/blob/trunk/crates/runtime/src/datafusion/mod.rs#L387